featurebyte/service/offline_store_feature_table_construction.py

```diff
@@ -0,0 +1,193 @@
+            offline_ingest_graphs=ingest_graph_metadata.offline_ingest_graphs,
+            source_type=source_type,
+        )
+
+        return OfflineStoreFeatureTableModel(
+            name=feature_table_name,
+            feature_ids=[feature.id for feature in features],
+            primary_entity_ids=[entity.id for entity in primary_entities],
+            # FIXME: consolidate all the offline store serving names (get_entity_id_to_serving_name_for_offline_store)
+            serving_names=[entity.serving_names[0] for entity in primary_entities],
+            feature_cluster=ingest_graph_metadata.feature_cluster,
+            output_column_names=ingest_graph_metadata.output_column_names,
+            output_dtypes=ingest_graph_metadata.output_dtypes,
+            entity_universe=entity_universe,
+            has_ttl=has_ttl,
+            feature_job_setting=feature_job_setting,
+        )
+
+    async def get_entity_universe_model(
+        self,
+        serving_names: List[str],
+        aggregate_result_table_names: List[str],
+        offline_ingest_graphs: List[OfflineStoreIngestQueryGraph],
+        source_type: SourceType,
+    ) -> EntityUniverseModel:
+        """
+        Create a new EntityUniverseModel object
+
+        Parameters
+        ----------
+        serving_names: List[str]
+            The serving names of the entities
+        aggregate_result_table_names: List[str]
+            The names of the aggregate result tables for window aggregates
+        offline_ingest_graphs: List[OfflineStoreIngestQueryGraph]
+            The offline ingest graphs that the entity universe is to be constructed from
+        source_type: SourceType
+            Source type information
+
+        Returns
+        -------
+        EntityUniverse
+        """
+        if aggregate_result_table_names:
+            universe_expr = construct_window_aggregates_universe(
+                serving_names=serving_names,
+                aggregate_result_table_names=aggregate_result_table_names,
+            )
+        else:
+            params = []
+            for offline_ingest_graph in offline_ingest_graphs:
+                primary_entities = [
+                    (await self.entity_service.get_document(entity_id))
+                    for entity_id in offline_ingest_graph.primary_entity_ids
+                ]
+                for info in offline_ingest_graph.aggregation_nodes_info:
+                    node = offline_ingest_graph.graph.get_node_by_name(info.node_name)
+                    # The entity universe has to be described in terms of the primary entity. If the
+                    # aggregation is based on a parent entity, we need to map it back to the primary
+                    # entity (a child).
+                    non_primary_entity_ids = self._get_non_primary_entity_ids(
+                        node,
+                        offline_ingest_graph.primary_entity_ids,
+                    )
```

Comment:
The `offline_ingest_graph.primary_entity_ids` here is w.r.t the offline store table. It may not be the primary entity of the feature/feature list. Just to double check whether this is expected or not.

Reply:
Yes, that is expected. The entity universe to be constructed here is also w.r.t to the offline store table.